Add metadata pattern docs (#578)
saleor/saleor-docs: docs/developer/extending/apps/apps-patterns/persistence-with-metadata-manager.mdx

---
title: App Pattern - Persistence with MetadataManager
---

## Why would you want a metadata manager?

Apps usually have configuration data that can be changed by the user during runtime:

- access keys to the external APIs (ex. key to a payment gateway)
- preferences like how the app should look (ex. customized messages)
- optional features which can be turned off or on (ex. if an additional notification should be sent to the user)

> Depending on the type of app you are creating, you might need a dedicated database for such data, but in most of cases, a simpler solution will be perfectly fine.

Reviewer: ```suggestion
Depending on the type of app you are creating, you might need a dedicated database for such data. However, in most of the cases, a simpler solution will be perfectly fine.
```

> Such a solution available out of the box is PrivateMetadata.

Reviewer: ```suggestion
Such a solution available out of the box is `PrivateMetadata`.
```

> Such a solution available out of the box is PrivateMetadata.

Author: Added 👍

> ## Whats Metadata and its features

Reviewer: ```suggestion
## What is Metadata and what are its features
```

> [Multiple of objects](developer/api-reference/interfaces/object-with-metadata.mdx) in the Saleor API has a way to add additional Metadata for the objects. In this document, we'll focus on App Metadata.

Reviewer 2: Many objects?

Reviewer 2: "Multiple of objects ... for the objects" - I think entire sentence can be rewritten

Author: Rephrased the sentence. Is it better now?

> `privateMetadata` field in the `app` object can be accessed only by the app itself or the users with `MANAGE_APPS` permission. The field contains a list of key and value pairs which can contain any string.

Reviewer 2: `...any string, including stringified json`

Author: Good idea. Added 👍

> `privateMetadata` field in the `app` object can be accessed only by the app itself or the users with `MANAGE_APPS` permission. The field contains a list of key and value pairs which can contain any string.

Reviewer: ```suggestion
`privateMetadata` field in the `app` object can be accessed only by the app itself or the users with `MANAGE_APPS` permission. The field contains a list of key and value pairs that can contain any string.
```

From the app developer's perspective there are a few pros:

- you don't need additional infrastructure to keep the data
- it plays well with the multi-tenant approach - each registered Saleor instance has its own preferences

> - when the app is uninstalled, the data are automatically removed

Reviewer: ```suggestion
- when the app is uninstalled, the data is automatically removed
```

## Using the Metadata

> You can access the API directly, or use utils provided by the SDK

Reviewer: ```suggestion
You can access the API directly, or use utils provided by the SDK.
```

### API

To get private metadata, run the query:

```graphql
query {
  app {
    privateMetadata {
      key
      value
    }
  }
}
```

:::note ID argument

> If you run the query with the app registration token, you don't need to specify the ID argument in the app query

Reviewer: ```suggestion
If you run the query with the app registration token, you don't need to specify the ID argument in the app query.
```

:::

Which will return:

```json
{
  "data": {
    "app": {
      "privateMetadata": [
        {
            "key": "Example key",
            "value": "example value"
        }
      ]
    }
  },
```

Setting new keys or updating the existing ones:

```graphql
mutation {
  updateMetadata(
    id: "QXBwOjE3OA=="
    input: [{ key: "new metadata", value: "example value" }]
  ) {
    errors {
      field
      message
    }
  }
}
```

### Metadata manager

> The App SDK comes with MetadataManager class which adds additional

Reviewer: ```
The App SDK comes with MetadataManager class which adds additional
```

I believe the sentence was finished prematurely. Adds additional what?

Author: Thats right, updated it 🙇

> - full documentation is available [here](https://github.com/saleor/saleor-app-sdk/blob/main/docs/settings-manager.md)

Reviewer 2: without "is"

"The App SDK comes with MetadataManager class which adds additional, full documentation, available here"

Author: Changed this section to be less confising. Used the notes section for that, WDYT?

- [example](https://github.com/saleor/saleor-app-template/tree/examples-metadata-manager) of how the Manager can be used in the App together with the interface

> Using this abstraction will give you a nicer interface to interact with and additional caching to improve performance.

Reviewer 2: "caching to improve performance and reduce API calls" ?

Entries in the manager are stored using the structure:

```
  key: string;
  value: string;
  domain?: string;
```

:::note Optional `domain` field

> Since metadata will be copied if you clone the database of Saleor instance, we recommend adding the parameter `domain` to prevent misusing secret keys from the different domains.  

Reviewer 2: "This can happend when Saleor Cloud promotes environment or database is cloned between Saleor instances" - maybe something like this, to give some context why this is a problem?

:::

Interacting with the API is done with the methods:

- `get: (key: string, domain?: string) => Promise<string | undefined>`
- `set: (settings: SettingsValue[] | SettingsValue) => Promise<void>`

#### Encrypted MetadataManager

> If you want to keep secrets, like non-public API keys we recommend `EncryptedMetadataManager` which will encrypt the values for the additional layer of security. `get` and `set` methods have the same interface

Reviewer: ```suggestion
If you want to keep secrets, like non-public API keys, we recommend using `EncryptedMetadataManager`. It will encrypt the values for the additional layer of security. `get` and `set` methods have the same interface.
```
